[google_maps_flutter] Implement polyline patterns in google maps ios

Hari-07 opened the pull request:
This PR Implements patterns in google maps ios polylines. Currently the patterns param is simply ignored on ios, despite the official google maps SDK for ios having instructions on how to achieve repeated patterns: https://developers.google.com/maps/documentation/ios-sdk/shapes#add-a-repeating-color-pattern-to-a-polyline

*List which issues are fixed by this PR. You must list at least one issue.*
https://github.com/flutter/flutter/issues/60083

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*
Nil

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

Hari-07 commented:
I don't currently have any tests, however that is because looking at the existing tests, since I only made changes to the ios package, it didn't seem right to author tests in the `google_maps_flutter` package. 

But looking at tests in `google_maps_flutter_ios` it seemed like they do not test things like what this PR covers

Open to authoring tests, if someone could let me know what the tests should test

stuartmorgan commented:
Marking as a draft for now since this needs native unit tests before it will be ready for review (per discussion in Discord).

Hari-07 commented:
Yeah just been on a break, should be done soon

Hari-07 commented:
I implemented tests for 2 out of 3 functions I wrote. For the 3rd function, the test should look something like 

```
- (void)testStrokeStylesFromPatterns {
    NSArray* patterns = @ [@[@"gap", @10], @[@"dash", @10]];
    UIColor* strokeColor = [UIColor redColor];
    
    NSArray<GMSStrokeStyle*> *patternStrokeStyle = [FLTGoogleMapJSONConversions strokeStylesFromPatterns:patterns strokeColor:strokeColor];
    
    XCTAssertEqual([patternStrokeStyle count], 2);
    
    GMSStrokeStyle* firstStrokeStyle = patternStrokeStyle[0];
    GMSStrokeStyle* secondStrokeStyle = patternStrokeStyle[1];
    
    // XCTAssertEqualObjects(firstStrokeStyle, [GMSStrokeStyle solidColor:[UIColor clearColor]]);
    // XCTAssertEqualObjects(secondStrokeStyle, [GMSStrokeStyle solidColor:strokeColor]);
}
```

However, the color property is not readable so I couldn't get this to work.

Hari-07 commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios11/ios/RunnerTests/GoogleMapsPolylinesControllerTests.m:
```diff
@@ -0,0 +1,52 @@
+- (void)testSetPatterns {
+    NSArray<GMSStrokeStyle *> *styles=@[[GMSStrokeStyle solidColor:[UIColor clearColor]], [GMSStrokeStyle solidColor:[UIColor redColor]]];
+    
+    NSArray<NSNumber *> *lengths= @[ @10, @10];
+    
+    XCTAssertNil(self.polylineController.polyline.spans);
+    
+    [self.polylineController setPattern:styles lengths:lengths];
+    
+    XCTAssertNotNil(self.polylineController.polyline.spans);
```
`spans` is an array of `GMSStrokeStyle` which doesn't have any inspectable values, so settled to making sure its not empty anymore

hellohuanlin commented on packages/google_maps_flutter/google_maps_flutter_android/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## NEXT
+
+* Adds support for patterns in polylines on iOS.
```
do we need this change in android?

Hari-07 replied:
I hadn't done this, and was getting some error from an automated test about README. So I ran the tool instead of doing the changelog manually, and that added this here

stuartmorgan replied:
You can revert this; we'll override the changelog check once it's ready to land. The tool errs on the side of adding changelogs for things so that wo don't just forget them, but the current `google_maps_flutter_android` change is uninteresting for clients, so shouldn't have an entry here.

stuartmorgan replied:
This still needs to be reverted.

hellohuanlin commented on packages/google_maps_flutter/google_maps_flutter/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## NEXT
```
this should be a bump

hellohuanlin commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios11/ios/RunnerTests/FLTGoogleMapJSONConversionsConversionTests.m:
```diff
@@ -287,4 +287,18 @@ - (void)testCameraUpdateFromChannelValueZoomTo {
 }
 
+- (void)testLengthsFromPatterns {
+  NSArray *patterns = @[ @[ @"gap", @10 ], @[ @"dash", @10 ] ];
```
can you use diff numbers for them?

hellohuanlin replied:
do they have to be ints?

Hari-07 replied:
Not necessarily. According to the docs it's an array of `NSNumber*` , so floats should be fine too

hellohuanlin replied:
can you also use floating values. just wanna make sure that works too

hellohuanlin commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios11/ios/RunnerTests/GoogleMapsPolylinesControllerTests.m:
```diff
@@ -0,0 +1,56 @@
+
+  [self.polylineController setPattern:styles lengths:lengths];
+
+  XCTAssertNotNil(self.polylineController.polyline.spans);
```
can you assert the actual values (rather than just asserting not nil)?

Hari-07 replied:
![image](https://github.com/flutter/packages/assets/22373191/785a10b3-be29-48b3-b4a9-fe374aed4126)


I think that's not possible. Spans is of type `NSArray<GMSStyleSpan *>`
But as you can see in the screenshot, a single span itself doesn't have any observable props

hellohuanlin replied:
does `isEqual` work?

Hari-07 replied:
I'm comparing just the first value of the array to illustrate the issue. I added a new `pattern` variable which should have the same value as that what's non nil



```
- (void)testSetPatterns {
  NSArray<GMSStrokeStyle *> *styles = @[
    [GMSStrokeStyle solidColor:[UIColor clearColor]], [GMSStrokeStyle solidColor:[UIColor redColor]]
  ];

  NSArray<NSNumber *> *lengths = @[ @10, @10 ];

  XCTAssertNil(self.polylineController.polyline.spans);

  [self.polylineController setPattern:styles lengths:lengths];

  NSArray<GMSStyleSpan *> *pattern =
      GMSStyleSpans(self.polylineController.polyline.path, styles, lengths, kGMSLengthRhumb);

//    XCTAssertNotNil(self.polylineController.polyline.spans);
  XCTAssertEqualObjects(self.polylineController.polyline.spans[0], pattern[0]);
}
```

On running this, I get an error that says
```
testSetPatterns: ((self.polylineController.polyline.spans[0]) equal to (pattern[0])) failed: ("<GMSStyleSpan: 0x600000a3ab80>") is not equal to ("<GMSStyleSpan: 0x600000b0fc60>")
```

<img width="381" alt="image" src="https://github.com/flutter/packages/assets/22373191/a6b79d2a-aa5c-4b13-8041-53dc7ec5c7d6">

<img width="397" alt="image" src="https://github.com/flutter/packages/assets/22373191/ed75716c-4618-4753-885e-649600fe31bf">
<br><br><br>
Looking at the segments in the debugger they do have the same value thought (Length of segment and color)
You can see that in the screenshots above and below. <br>
<img width="392" alt="image" src="https://github.com/flutter/packages/assets/22373191/7193cb01-1c46-4051-8a25-71e8fcd13b13">

<img width="446" alt="image" src="https://github.com/flutter/packages/assets/22373191/33e9caaa-64ac-45b1-9789-fb9c84ab4def">

stuartmorgan replied:
The default implementation of `isEqual:` is just pointer equality, so it's not surprising that it doesn't work here; GMS would have had to explicitly implement it for this object.

Hari-07 replied:
So just stick to checking non nil for now?

hellohuanlin replied:
If it's not possible to check the actual value, nothing we can do. Can you add a comment there explaining this?

Hari-07 replied:
Alright done!

hellohuanlin commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FLTGoogleMapJSONConversions.m:
```diff
@@ -141,4 +141,34 @@ + (nullable GMSCameraUpdate *)cameraUpdateFromChannelValue:(NSArray *)channelVal
 }
+
++ (NSArray<GMSStrokeStyle *> *)strokeStylesFromPatterns:(NSArray *)patterns
+                                            strokeColor:(UIColor *)strokeColor {
```
is this function tested?

Hari-07 replied:
Had left a [comment](https://github.com/flutter/packages/pull/5757#issuecomment-1876560460) above about why I think this cant be tested

Hari-07 replied:
@hellohuanlin Is there any next step on this?

stuartmorgan replied:
It would be good to have a test that you get back an array with the expected number of elements (and a comment explaining why that's all it's testing). It should include both gap and non-gap elements. That will at least exercise all of the codepaths here, so that, e.g., if one of them had a bug that caused an exception, we'd catch that.

Hari-07 replied:
Alright!

jmagman commented:
@Hari-07 can you fix up the conflicting files, including the CHANGELOG and pubspec versions?  Then this will be ready to merge (unless @stuartmorgan has further comments)

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 2.5.3
```
There should not be a changelog entry in this package; nothing relevant to publishing has changed.

Hari-07 replied:
> this should be a bump

stuartmorgan replied:
> How would the ios dependency change if there's no version change here?

I'm not sure what you are asking here; the iOS implementation is a completely separate package.

> Or you mean version bump and no changelog, isn't that kinda weird?

No, a package whose only change is a minor adjustment to an unpublished example file should have no version change or CHANGELOG entry.

Hari-07 replied:
>I'm not sure what you are asking here; the iOS implementation is a completely separate package.

What I meant was that `google_maps_flutter_ios` will have a new version right, so how would users get this new version, since the dependency on it is through `google_maps_flutter`

stuartmorgan replied:
The same way they would get updates to any transitive dependencies; they would run `flutter pub upgrade`.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 2.3.6
+
+* Adds support for patterns in polylines on iOS.
```
Nit: Remove "on iOS", since this is the iOS package.

stuartmorgan replied:
This still says "on iOS". Please make sure you are not marking things as resolved when the requested change has not been made, as it makes review much harder.

Hari-07 replied:
Agreed 100%. Definitely resolved it accidentally, have made that change now

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 2.3.6
```
Nit: 2.4.0 since this is a new feature.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios11/ios/RunnerTests/GoogleMapsPolylinesControllerTests.m:
```diff
@@ -0,0 +1,57 @@
+#import "PartiallyMockedMapView.h"
+
+@interface GoogleMapsPolylinesControllerTests : XCTestCase
+@property(strong, atomic) FLTGoogleMapPolylineController *polylineController;
```
Fixtures should not be stateful without a very compelling reason; this can just be a local variable in each method, and `setUp` can become a utility method that creates and returns it (e.g., `controllerWithMockedMap`).

Hari-07 replied:
Wasn't really sure how to share state between `setUp` and other functions without making them stateful. Just made it a kind of helper function instead of the overriden setup. The annoying part is that now other tests must call this function manually.

stuartmorgan replied:
> The annoying part is that now other tests must call this function manually.

That is a feature, as it makes interdependencies explicit and local. See https://abseil.io/tips/122

Hari-07 replied:
That makes a lot of sense thinking more about. 🔥

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios11/ios/RunnerTests/GoogleMapsPolylinesControllerTests.m:
```diff
@@ -0,0 +1,57 @@
+
+  [self.polylineController setPattern:styles lengths:lengths];
+
+  // `GMSStyleSpan` doesn't implement `isEqual` so cannot be compared by value at present
```
Nit: this is missing a period.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FLTGoogleMapJSONConversions.h:
```diff
@@ -24,6 +24,9 @@ NS_ASSUME_NONNULL_BEGIN
 + (nullable GMSCameraUpdate *)cameraUpdateFromChannelValue:(NSArray *)channelValue;
++ (NSArray<GMSStrokeStyle *> *)strokeStylesFromPatterns:(NSArray *)patterns
+                                            strokeColor:(UIColor *)strokeColor;
++ (NSArray<NSNumber *> *)spanLengthsFromPatterns:(NSArray *)patterns;
```
These need declaration comments, per the Obj-C style guide. The existing methods violate the guide, which is problematic, but that shouldn't be perpetuated in new methods.

stuartmorgan replied:
This has not been addressed.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapPolylineController.h:
```diff
@@ -7,10 +7,13 @@
 
 // Defines polyline controllable by Flutter.
 @interface FLTGoogleMapPolylineController : NSObject
+@property(strong, nonatomic) GMSPolyline *polyline;
```
This needs a documentation comment, per style guide.

Why is it public? If it's just for testing, it should be in a `GoogleMapPolylineControlller_Test.h` instead.

Hari-07 replied:
Yeah, great call

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapPolylineController.h:
```diff
@@ -7,10 +7,13 @@
                           identifier:(NSString *)identifier
                              mapView:(GMSMapView *)mapView;
 - (void)removePolyline;
+- (void)setPattern:(NSArray<GMSStrokeStyle *> *)styles lengths:(NSArray<NSNumber *> *)lengths;
```
This needs a declaration comment.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapPolylineController.h:
```diff
@@ -22,4 +25,5 @@
 - (void)removePolylineWithIdentifiers:(NSArray *)identifiers;
 - (void)didTapPolylineWithIdentifier:(NSString *)identifier;
 - (bool)hasPolylineWithIdentifier:(NSString *)identifier;
++ (GMSMutablePath *)getPath:(NSDictionary *)polyline;
```
As does this.

stuartmorgan replied:
Obj-C methods do not start with `get`, and they should name their parameters in the signature. This should be called `pathForPolyline`.

stuartmorgan replied:
Why is this public?

Hari-07 commented:
@stuartmorgan Thanks for the comprehensive review. I believe I've resolved all of them. Hope to get this merged in soon

jmagman commented:
@stuartmorgan any more changes requested here?  Or is this ready to be merged (pending updating the CHANGELOG and pubspec again)?

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios12/ios/RunnerTests/GoogleMapPolylineController_Test.h:
```diff
@@ -0,0 +1,13 @@
+
+#import "GoogleMapPolylineController.h"
+
+@interface FLTGoogleMapPolylineController (Test)
```
Needs a declaration comment, per style guide. Something like "Private APIs exposed for unit testing."

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios12/ios/RunnerTests/GoogleMapPolylineController_Test.h:
```diff
@@ -0,0 +1,13 @@
+#import "GoogleMapPolylineController.h"
+
+@interface FLTGoogleMapPolylineController (Test)
+@property(strong, nonatomic) GMSPolyline *polyline;
```
Needs a declaration comment, per style guide.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios12/ios/RunnerTests/GoogleMapPolylineController_Test.h:
```diff
@@ -0,0 +1,13 @@
+@property(strong, nonatomic) GMSPolyline *polyline;
+@end
+
+@interface FLTPolylinesController (Test)
```
Same.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios12/ios/RunnerTests/GoogleMapPolylineController_Test.h:
```diff
@@ -0,0 +1,13 @@
+@end
+
+@interface FLTPolylinesController (Test)
+- (GMSMutablePath *)pathForPolyline:(NSDictionary *)polyline;
```
Same.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FLTGoogleMapJSONConversions.m:
```diff
@@ -141,4 +141,34 @@ + (nullable GMSCameraUpdate *)cameraUpdateFromChannelValue:(NSArray *)channelVal
+      strokeStyle = [GMSStrokeStyle solidColor:strokeColor];
+    }
+
+    [strokeStyles addObject:strokeStyle];
```
Consider collapsing this all to:
```objc
UIColor* color = [patternType isEqualToString:@"gap"] ? [UIColor clearColor] : strokeColor;
[strokeStyles addObject:[GMSStrokeStyle solidColor:color];
```

Having mostly identical code in both branches of a conditional is generally a code smell, since it makes it easy for things to accidentally diverge.

Hari-07 replied:
Great call!

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FLTGoogleMapJSONConversions.m:
```diff
@@ -141,4 +141,34 @@ + (nullable GMSCameraUpdate *)cameraUpdateFromChannelValue:(NSArray *)channelVal
   return nil;
 }
+
++ (NSArray<GMSStrokeStyle *> *)strokeStylesFromPatterns:(NSArray *)patterns
```
The type of `patterns` should specify its lightweight generic.

Hari-07 replied:
I didn't quite get what you mean, I assume you want me to specify the type somehow? But it's just a json object where the first member is a string (the pattern type) and second a number (the length)

And it's not a dictionary

stuartmorgan replied:
> I didn't quite get what you mean, I assume you want me to specify the type somehow?

Yes, using the lightweight generics syntax, as is done for the return value.

> But it's just a json object where the first member is a string (the pattern type) and second a number (the length)
> 
> And it's not a dictionary

I didn't say it was a dictionary, I just said its type needs to be specified.

> Plus the same has been followed in existing code, for example
> 
> ```
> + (GMSCoordinateBounds *)coordinateBoundsFromLatLongs:(NSArray *)latlongs {
>   return [[GMSCoordinateBounds alloc]
>       initWithCoordinate:[FLTGoogleMapJSONConversions locationFromLatLong:latlongs[0]]
>               coordinate:[FLTGoogleMapJSONConversions locationFromLatLong:latlongs[1]]];
> }
> ```

The existence of older code that [does not follow the current style guide](https://google.github.io/styleguide/objcguide.html#use-lightweight-generics-to-document-contained-types) does not mean that new code doesn't have to follow it.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FLTGoogleMapJSONConversions.m:
```diff
@@ -141,4 +141,34 @@ + (nullable GMSCameraUpdate *)cameraUpdateFromChannelValue:(NSArray *)channelVal
+  return strokeStyles;
+}
+
++ (NSArray<NSNumber *> *)spanLengthsFromPatterns:(NSArray *)patterns {
```
Same here.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FLTGoogleMapJSONConversions.m:
```diff
@@ -141,4 +141,34 @@ + (nullable GMSCameraUpdate *)cameraUpdateFromChannelValue:(NSArray *)channelVal
+
++ (NSArray<GMSStrokeStyle *> *)strokeStylesFromPatterns:(NSArray *)patterns
+                                            strokeColor:(UIColor *)strokeColor {
+  NSMutableArray *strokeStyles = [[NSMutableArray alloc] init];
```
Nit: use `initWithCapacity:` since you know the length in advance, as it's more efficient.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FLTGoogleMapJSONConversions.m:
```diff
@@ -141,4 +141,34 @@ + (nullable GMSCameraUpdate *)cameraUpdateFromChannelValue:(NSArray *)channelVal
+}
+
++ (NSArray<NSNumber *> *)spanLengthsFromPatterns:(NSArray *)patterns {
+  NSMutableArray *lengths = [[NSMutableArray alloc] init];
```
`initWithCapacity:` here as well.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapPolylineController.m:
```diff
@@ -95,6 +99,14 @@ - (void)interpretPolylineOptions:(NSDictionary *)data
   }
+
+  NSArray *patterns = data[@"pattern"];
+  if (patterns && patterns != (id)[NSNull null]) {
```
All of the code above here is incorrect, and this shouldn't replicate that incorrect pattern; assigning a `NSNull*` to a variable of type `NSArray*` is a type violation. The cast to `(id)` here is a workaround for the compiler calling out that problem.

The better way to handle this would be to make a static helper function:
```objc
/**
 * Returns dict[key], or nil if dict[key] is NSNull.
 */
id GetValueOrNil(NSDictionary *dict, NSString *key) {
  id value = dict[key];
  return value == [NSNull null] ? nil : value;
}
```

The this code becomes:
```objc
NSArray *patterns = GetValueOrNil(data, @"pattern");
if (patterns) {
  ...
```

(If you want to update the code above to use the helper too, that would be great.)

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapPolylineController.m:
```diff
@@ -106,6 +118,14 @@ @interface FLTPolylinesController ()
+ *
+ * @param polyline The polyline instance for which path is calculated.
+ * @return An instance of GMSMutablePath.
+ */
```
This should go on the declaration in the test header.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapPolylineController.m:
```diff
@@ -106,6 +118,14 @@ @interface FLTPolylinesController ()
+ * @param polyline The polyline instance for which path is calculated.
+ * @return An instance of GMSMutablePath.
+ */
+- (GMSMutablePath *)pathForPolyline:(NSDictionary *)polyline;
```
This should be removed, as it's redundant with the test header.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapPolylineController.m:
```diff
@@ -171,7 +191,7 @@ - (bool)hasPolylineWithIdentifier:(NSString *)identifier {
   return self.polylineIdentifierToController[identifier] != nil;
 }
-+ (GMSMutablePath *)getPath:(NSDictionary *)polyline {
+- (GMSMutablePath *)pathForPolyline:(NSDictionary *)polyline {
```
Why was this changed to a non-static method?

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapPolylineController.m:
```diff
@@ -171,7 +191,7 @@ - (bool)hasPolylineWithIdentifier:(NSString *)identifier {
-+ (GMSMutablePath *)getPath:(NSDictionary *)polyline {
+- (GMSMutablePath *)pathForPolyline:(NSDictionary *)polyline {
   NSArray *pointArray = polyline[@"points"];
   NSArray<CLLocation *> *points = [FLTGoogleMapJSONConversions pointsFromLatLongs:pointArray];
```
Actually, if this is just doing a JSON conversion, is there any reason not to just move this to FLTGoogleMapJSONConversions.m/.h, and eliminate the need to expose a private method at all?

stuartmorgan replied:
I don't see a response or change related to this comment.

Hari-07 replied:
I just realised those comments never got posted. It said "Pending" and was stuck like that. Tripped up by github

<img width="791" alt="image" src="https://github.com/flutter/packages/assets/22373191/aa848278-1b37-4544-8337-ba02f639b4b0">

stuartmorgan replied:
> I'd assume the FLTJSONConversion file ought to be dumb and just convert and needn't know about the structure of the polyline object (`polyline[@"points"]`)

I'm not sure what this means; what would it mean for a JSON conversion method to not understand the structure of the object it is converting?

Hari-07 replied:
As in this function calls through to a JSON conversion method (`pointsFromLatLongs`).

Other than that its all dealing with objects of maps, not untyped json fields. So I thought it makes more sense to be a method within the class. I have currently made it static again. Im not sure if this ought to be static/non-static but assumed you preferred it to be static based on above comments.

stuartmorgan replied:
> Other than that its all dealing with objects of maps, not untyped json fields.

Maps are a core JSON type, so I'm still not clear on what the distinction is here. But it's fine to leave it here for now.

stuartmorgan commented:
From triage: @Hari-07 Are you still planning on updating this PR to address the review comments above?

Hari-07 commented:
> From triage: @Hari-07 Are you still planning on updating this PR to address the review comments above?

Kinda slipped my mind totally, will do it over this weekend

vashworth commented:
@Hari-07 Is this ready for re-review?

Hari-07 commented:
> @Hari-07 Is this ready for re-review?

Yeah it should be good to go, I had a couple of questions that I've commented above. I believe I've addressed all the comments @stuartmorgan had made

yaakovschectman reviewed (approved):
LGTM for Android changes

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios12/ios/RunnerTests/GoogleMapPolylineController_Test.h:
```diff
@@ -0,0 +1,29 @@
+
+/**
+ *  Internal APIs exposed for unit testing
+ */
```
We've recently standardized on `///` for declaration comments in Obj-C code, so these should all be updated to use that.

(Applies throughout the PR.)

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios12/ios/RunnerTests/GoogleMapPolylineController_Test.h:
```diff
@@ -0,0 +1,29 @@
+ */
+@interface FLTPolylinesController (Test)
+/**
+ * Returns the path for polyline based on the points(locations) the polyline has.
```
Missing space before `(`.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios12/ios/RunnerTests/GoogleMapPolylineController_Test.h:
`_Test` headers go next to the actual headers, not in the test directory.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios12/ios/RunnerTests/GoogleMapsPolylinesControllerTests.m:
```diff
@@ -0,0 +1,80 @@
+
+@implementation GoogleMapsPolylinesControllerTests
+
+- (FLTGoogleMapPolylineController *)setUpPolyLineControllerWithMockedMap {
```
This needs a declaration comment, per style guide. Test helper code is still code that humans have to understand and maintain, which is what comments are to help with.

stuartmorgan replied:
This should just be called `polylineControllerWithMockedMap` since it returns one; `setUp...` doesn't sound like a method that returns anything.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FLTGoogleMapJSONConversions.m:
```diff
@@ -141,4 +141,28 @@ + (nullable GMSCameraUpdate *)cameraUpdateFromChannelValue:(NSArray *)channelVal
+
++ (NSArray<NSNumber *> *)spanLengthsFromPatterns:(NSArray<NSArray *> *)patterns {
+  NSMutableArray *lengths = [[NSMutableArray alloc] initWithCapacity:[patterns count]];
+  for (unsigned int i = 0; i < [patterns count]; i++) {
```
This should use fast enumeration rather than indexing, and `addObject:` to `lengths`.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/FLTGoogleMapJSONConversions.m:
```diff
@@ -141,4 +141,28 @@ + (nullable GMSCameraUpdate *)cameraUpdateFromChannelValue:(NSArray *)channelVal
++ (NSArray<GMSStrokeStyle *> *)strokeStylesFromPatterns:(NSArray<NSArray *> *)patterns
+                                            strokeColor:(UIColor *)strokeColor {
+  NSMutableArray *strokeStyles = [[NSMutableArray alloc] initWithCapacity:[patterns count]];
+  for (unsigned int i = 0; i < [patterns count]; i++) {
```
This should use fast enumeration rather than indexing, and `addObject:` to `strokeStyles`.

stuartmorgan reviewed (changes requested):
We've collapsed the iOS 12 example into the iOS 14 example due to dropping iOS 12 and iOS 13 support, so you'll need to migrate the test changes to that example. Sorry for the churn!

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapPolylineController.m:
```diff
@@ -12,6 +12,15 @@ @interface FLTGoogleMapPolylineController ()
+/**
+ * Returns dict[key], or nil if dict[key] is NSNull.
+ */
+id GetValueOrNilFromDict(NSDictionary * dict, NSString *key) {
+    id value = dict[key];
+    return value == [NSNull null] ? nil : value;
+}
```
It's fine to leave this in the file that needs it, although it should be declared `static`.

(It will go way soon, when we convert to Pigeon, so making a new file will just add extra work to undo later.)

stuartmorgan commented:
> I could also specify the internal array as an NSObject and then cast it explicitly. Not sure if that's any better though

For some reason GitHub won't let me respond inline to this, but yes, that would be better. It makes it explicit in the code that the array is heterogenous, which is unusual and thus better to have be as explicit as possible.

stuartmorgan commented:
@Hari-07 Is this ready for review again?

Hari-07 commented:
I've updated for all your comments other than [this thread](https://github.com/flutter/packages/pull/5757#discussion_r1492837110). Have left a comment on that too.

Apologies about the delay

Hari-07 commented:
One of the test failure is that the new _Test header isn't in the umbrella header. Do we need to put test headers into it tho?

stuartmorgan commented:
> One of the test failure is that the new _Test header isn't in the umbrella header. Do we need to put test headers into it tho?

Yes, unfortunately the way our builds are structured we have to make the test headers public currently. (Longer term, moving to Swift will resolve this problem.)

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios14/ios/RunnerTests/GoogleMapsPolylinesControllerTests.m:
```diff
@@ -0,0 +1,83 @@
+///  @return An object of FLTGoogleMapPolylineController
+///
+- (FLTGoogleMapPolylineController *)polylineControllerWithMockedMap {
+  NSDictionary *polyline = @{
```
`NSDictionary<NSString*, id>`

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios14/ios/RunnerTests/GoogleMapsPolylinesControllerTests.m:
```diff
@@ -0,0 +1,83 @@
+}
+
+- (void)testStrokeStylesFromPatterns {
+  NSArray *patterns = @[ @[ @"gap", @10 ], @[ @"dash", @10 ] ];
```
`NSArray<NSArray<id> *>`

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/example/ios14/ios/RunnerTests/FLTGoogleMapJSONConversionsConversionTests.m:
```diff
@@ -288,4 +288,18 @@ - (void)testCameraUpdateFromChannelValueZoomTo {
 }
 
+- (void)testLengthsFromPatterns {
+  NSArray *patterns = @[ @[ @"gap", @10 ], @[ @"dash", @6.4 ] ];
```
`NSArray<NSArray<id> *>`

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapPolylineController.h:
```diff
@@ -11,6 +11,14 @@
                              mapView:(GMSMapView *)mapView;
 - (void)removePolyline;
+
+///
```
No blank comment lines at the start and end of the comment please.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapPolylineController.m:
```diff
@@ -12,6 +12,14 @@ @interface FLTGoogleMapPolylineController ()
 
 @end
 
+///
```
Same here.

stuartmorgan commented on packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapPolylineController_Test.h:
```diff
@@ -0,0 +1,29 @@
+
+#import "GoogleMapPolylineController.h"
+
+///
```
Same note for all of these comments.

stuartmorgan reviewed (changes requested):
Just minor nits, other than that, LGTM once it passes CI!

Hari-07 commented:
@stuartmorgan Thank you so much for the detailed reviews, and the back and forth. This was a **long** one. I didn't have prior experience with Obj-C and was figuring it out as we went hence the many beginner mistakes. 

Have fixed your latest round of suggestions as well, the only CI failure is its expecting a changelog change in `google_maps_flutter` but as per this [conversation](https://github.com/flutter/packages/pull/5757#discussion_r1474601297) that's not required.

Hope this can get merged in soon 🙏

stuartmorgan commented:
CHANGELOG override: updating non-main examples to not check for iOS in other packages doesn't warrant client-facing commenting.

stuartmorgan reviewed (approved):
LGTM, thanks!
